pkg/bloomgateway/sharding.go

```diff
@@ -0,0 +1,162 @@
+		limits:       l,
+		instanceAddr: instanceAddr,
+		instanceID:   instanceID,
+		logger:       logger,
+	}
+}
+
+// FilterTenants implements ShardingStrategy.
+func (s *ShuffleShardingStrategy) FilterTenants(_ context.Context, tenantIDs []string) ([]string, error) {
+	// As a protection, ensure the bloom gateway instance is healthy in the ring. It could also be missing
+	// in the ring if it was failing to heartbeat the ring and it got remove from another healthy bloom gateway
+	// instance, because of the auto-forget feature.
+	if set, err := s.r.GetAllHealthy(BlocksOwnerSync); err != nil {
+		return nil, err
+	} else if !set.Includes(s.instanceAddr) {
+		return nil, errGatewayUnhealthy
+	}
+
+	var filteredIDs []string
+
+	for _, tenantID := range tenantIDs {
+		subRing := GetShuffleShardingSubring(s.r, tenantID, s.limits)
+
+		// Include the user only if it belongs to this bloom gateway shard.
+		if subRing.HasInstance(s.instanceID) {
+			filteredIDs = append(filteredIDs, tenantID)
+		}
+	}
+
+	return filteredIDs, nil
+}
+
+func getBucket(rangeMin, rangeMax, pos uint64) int {
+	return 0
+}
+
+// FilterBlocks implements ShardingStrategy.
+func (s *ShuffleShardingStrategy) FilterBlocks(ctx context.Context, tenantID string, blockRefs []BlockRef) ([]BlockRef, error) {
+	filteredBlockRefs := make([]BlockRef, 0, len(blockRefs))
+
+	subRing := GetShuffleShardingSubring(s.r, tenantID, s.limits)
+
+	bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+	var rs ring.ReplicationSet
+	var err error
+
+	for _, blockRef := range blockRefs {
+		rs, err = subRing.Get(uint32(blockRef.FromFp), BlocksOwnerSync, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			return nil, err
+		}
+		// Include the block only if it belongs to this bloom gateway shard.
+		if rs.Includes(s.instanceID) {
+			filteredBlockRefs = append(filteredBlockRefs, blockRef)
+			continue
+		}
+
+		rs, err = subRing.Get(uint32(blockRef.ThroughFp), BlocksOwnerSync, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			return nil, err
+		}
+		// Include the block only if it belongs to this bloom gateway shard.
+		if rs.Includes(s.instanceID) {
+			filteredBlockRefs = append(filteredBlockRefs, blockRef)
+			continue
+		}
+	}
+	return filteredBlockRefs, nil
+}
+
+// GetShuffleShardingSubring returns the subring to be used for a given user.
+// This function should be used both by index gateway servers and clients in
+// order to guarantee the same logic is used.
+func GetShuffleShardingSubring(ring ring.ReadRing, tenantID string, limits Limits) ring.ReadRing {
+	shardSize := limits.BloomGatewayShardSize(tenantID)
+
+	// A shard size of 0 means shuffle sharding is disabled for this specific user,
+	// so we just return the full ring so that indexes will be sharded across all index gateways.
+	// Since we set the shard size to replication factor if shard size is 0, this
+	// can only happen if both the shard size and the replication factor are set
+	// to 0.
+	if shardSize <= 0 {
```

Comment:
can shardSize be negative? why not `==`

Reply:
`shardSize` is an `int`. And the config would allow setting a negative value.